File: python/mirage/vfs/airtable/airtable.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any

from mirage.accessor.airtable import AirtableAccessor
from mirage.commands.builtin.airtable import COMMANDS
from mirage.commands.builtin.airtable.io import IO
from mirage.core.airtable.config import AirtableConfig
from mirage.ops.airtable import OPS
from mirage.types import VFSName
from mirage.vfs.airtable.prompt import PROMPT
from mirage.vfs.bound import BoundVFS


class AirtableVFS(BoundVFS):
    """Airtable bases as directories, tables as records.jsonl files.

    Records are live data another client may edit at any moment, so reads
    are never served from the file cache; the schema listings still ride
    the index for its TTL.

    Args:
        config (AirtableConfig): the account and its bounds.
    """

    accessor: AirtableAccessor
    name: str = VFSName.AIRTABLE
    caches_reads: bool = False
    PROMPT: str = PROMPT

    def __init__(self, config: AirtableConfig) -> None:
        super().__init__(io=IO)
        self.config = config
        self.accessor = AirtableAccessor(self.config)
        for fn in COMMANDS:
            self.register(fn)
        for op in OPS:
            self.register_op(op)

    def get_state(self) -> dict[str, Any]:
        return self.config_state(self.config)

File: typescript/packages/core/src/vfs/airtable/airtable.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { AirtableAccessor } from '../../accessor/airtable.ts'
import { AIRTABLE_COMMANDS } from '../../commands/builtin/airtable/index.ts'
import { makeResolveGlob } from '../../commands/builtin/generic_bind/index.ts'
import type { RegisteredCommand } from '../../commands/config.ts'
import {
  redactAirtableConfig,
  type AirtableConfig,
  type AirtableConfigRedacted,
} from '../../core/airtable/config.ts'
import { read } from '../../core/airtable/read.ts'
import { readdir } from '../../core/airtable/readdir.ts'
import { stat } from '../../core/airtable/stat.ts'
import { AIRTABLE_OPS } from '../../ops/airtable/index.ts'
import type { RegisteredOp } from '../../ops/registry.ts'
import { VFSName, type FileStat, type PathSpec } from '../../types.ts'
import { BaseVFS, type VFS } from '../base.ts'
import { AIRTABLE_PROMPT, AIRTABLE_WRITE_PROMPT } from './prompt.ts'

const resolveGlob = makeResolveGlob(readdir)

export interface AirtableVFSState {
  type: string
  config: AirtableConfigRedacted
}

/**
 * Airtable bases as directories, tables as records.jsonl files. Records are
 * live data another client may edit at any moment, so reads are never served
 * from the file cache; the schema listings still ride the index for its TTL.
 * The transport is plain fetch, so one class serves node and the browser.
 */
export class AirtableVFS extends BaseVFS implements VFS {
  readonly kind: string = VFSName.AIRTABLE
  readonly cachesReads: boolean = false
  // records.jsonl and the view files render a paged read, so their size is
  // unknown until the bytes exist.
  readonly sizesAlwaysKnown: boolean = false
  readonly supportsSnapshot: boolean = false
  readonly prompt: string = AIRTABLE_PROMPT
  readonly writePrompt: string = AIRTABLE_WRITE_PROMPT
  readonly accessor: AirtableAccessor

  private readonly config: AirtableConfig

  constructor(config: AirtableConfig, options: { fetchFn?: typeof fetch } = {}) {
    super()
    this.config = config
    this.accessor = new AirtableAccessor(config, options)
  }

  open(): Promise<void> {
    return Promise.resolve()
  }

  commands(): readonly RegisteredCommand[] {
    return AIRTABLE_COMMANDS
  }

  ops(): readonly RegisteredOp[] {
    return AIRTABLE_OPS
  }

  glob(paths: readonly PathSpec[], _prefix = ''): Promise<PathSpec[]> {
    return resolveGlob(this.accessor, paths, this.index)
  }

  readFile(path: PathSpec): Promise<Uint8Array> {
    return read(this.accessor, path, this.index)
  }

  readdir(path: PathSpec): Promise<string[]> {
    return readdir(this.accessor, path, this.index)
  }

  stat(path: PathSpec): Promise<FileStat> {
    return stat(this.accessor, path, this.index)
  }

  override getState(): AirtableVFSState {
    return { type: this.kind, config: redactAirtableConfig(this.config) }
  }

  override loadState(_state: AirtableVFSState): Promise<void> {
    return Promise.resolve()
  }
}
